src/contracts/exchange/NiftyswapExchange20.sol

```diff
@@ -0,0 +1,810 @@
+      currencyReserves[idBought] = currencyReserve.add(currencyAmount);
+    }
+
+    // Refund currency token if any
+    if (totalRefundCurrency > 0) {
+      TransferHelper.safeTransfer(currency, _recipient, totalRefundCurrency);
+    }
+
+    // Send Tokens all tokens purchased
+    token.safeBatchTransferFrom(address(this), _recipient, _tokenIds, _tokensBoughtAmounts, "");
+    return currencySold;
+  }
+
+  /**
+   * @dev Pricing function used for converting between currency token to Tokens.
+   * @param _assetBoughtAmount  Amount of Tokens being bought.
+   * @param _assetSoldReserve   Amount of currency tokens in exchange reserves.
+   * @param _assetBoughtReserve Amount of Tokens (output type) in exchange reserves.
+   * @return price Amount of currency tokens to send to Niftyswap.
+   */
+  function getBuyPrice(
+    uint256 _assetBoughtAmount,
+    uint256 _assetSoldReserve,
+    uint256 _assetBoughtReserve)
+    override public pure returns (uint256 price)
+  {
+    // Reserves must not be empty
+    require(_assetSoldReserve > 0 && _assetBoughtReserve > 0, "NiftyswapExchange20#getBuyPrice: EMPTY_RESERVE");
+
+    // Calculate price with fee
+    uint256 numerator = _assetSoldReserve.mul(_assetBoughtAmount).mul(1000);
+    uint256 denominator = (_assetBoughtReserve.sub(_assetBoughtAmount)).mul(FEE_MULTIPLIER);
+    (price, ) = divRound(numerator, denominator);
+    return price; // Will add 1 if rounding error
+  }
+
+  /**
+   * @notice Convert Tokens _id to currency tokens and transfers Tokens to recipient.
+   * @dev User specifies EXACT Tokens _id sold and MINIMUM currency tokens received.
+   * @dev Assumes that all trades will be valid, or the whole tx will fail
+   * @dev Sorting _tokenIds is mandatory for efficient way of preventing duplicated IDs (which would lead to errors)
+   * @param _tokenIds          Array of Token IDs that are sold
+   * @param _tokensSoldAmounts Array of Amount of Tokens sold for each id in _tokenIds.
+   * @param _minCurrency       Minimum amount of currency tokens to receive
+   * @param _deadline          Timestamp after which this transaction will be reverted
+   * @param _recipient         The address that receives output currency tokens.
+   * @return currencyBought How much currency was actually purchased.
+   */
+  function _tokenToCurrency(
+    uint256[] memory _tokenIds,
+    uint256[] memory _tokensSoldAmounts,
+    uint256 _minCurrency,
+    uint256 _deadline,
+    address _recipient)
+    internal nonReentrant() returns (uint256[] memory currencyBought)
+  {
+    // Number of Token IDs to deposit
+    uint256 nTokens = _tokenIds.length;
+
+    // Input validation
+    require(_deadline >= block.timestamp, "NiftyswapExchange20#_tokenToCurrency: DEADLINE_EXCEEDED");
+
+    // Initialize variables
+    uint256 totalCurrency = 0; // Total amount of currency tokens to transfer
+    currencyBought = new uint256[](nTokens);
+    uint256[] memory tokenReserves = new uint256[](nTokens);
```

Comment:
Same as before, I don't think this array has to be declared in a different line.

Reply:
Will update